File: packages/go/rules_watcher.go
package autoblock

import (
	"context"
	"encoding/json"
	"log/slog"
	"sync/atomic"
	"time"
)

const defaultPollInterval = 30 * time.Second

// RulesWatcher polls ab:{tenant}:rules:endpoint every interval and hot-reloads
// the Limiter's rule set without restarting the process.
//
// Usage:
//
//	watcher := limiter.NewWatcher(autoblock.WatcherOptions{})
//	go watcher.Run(ctx)
type RulesWatcher struct {
	limiter  *Limiter
	interval time.Duration
	onReload func([]Rule)
	onError  func(error)
}

// WatcherOptions configure hot-reload behaviour.
type WatcherOptions struct {
	// Interval between Redis polls. Default: 30 s.
	Interval time.Duration
	// OnReload is called after each successful rule swap.
	OnReload func([]Rule)
	// OnError is called when a poll cycle fails.
	OnError func(error)
}

// NewWatcher creates a RulesWatcher attached to this Limiter.
func (l *Limiter) NewWatcher(opts WatcherOptions) *RulesWatcher {
	interval := opts.Interval
	if interval <= 0 {
		interval = defaultPollInterval
	}
	return &RulesWatcher{
		limiter:  l,
		interval: interval,
		onReload: opts.OnReload,
		onError:  opts.OnError,
	}
}

// Run blocks, polls Redis on a ticker, AND subscribes to the
// ab:{tenant}:rules:changed pub/sub channel for instant invalidation.
// Intended to be launched in a goroutine: go watcher.Run(ctx)
func (w *RulesWatcher) Run(ctx context.Context) {
	// Immediate first load before the ticker fires.
	if err := w.Poll(ctx); err != nil {
		w.handleError(err)
	}

	// Subscribe to push notifications from the engine.
	// On any message we poll immediately instead of waiting for the ticker.
	pushCh := w.subscribePush(ctx)

	ticker := time.NewTicker(w.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := w.Poll(ctx); err != nil {
				w.handleError(err)
			}
		case _, ok := <-pushCh:
			if !ok {
				// Channel closed — subscription dropped; fall back to ticker only.
				pushCh = nil
				continue
			}
			slog.Debug("autoblock RulesWatcher: push notification received, reloading immediately")
			if err := w.Poll(ctx); err != nil {
				w.handleError(err)
			}
		}
	}
}

// subscribePush subscribes to the rules:changed pub/sub channel and returns a
// channel that receives one value per publish event.  Returns a nil channel if
// the subscription fails (falls back gracefully to ticker-only mode).
func (w *RulesWatcher) subscribePush(ctx context.Context) <-chan struct{} {
	channel := w.limiter.keys.rulesChanged()
	pubsub := w.limiter.cfg.Redis.Subscribe(ctx, channel)

	// Verify subscription is alive.
	if _, err := pubsub.Receive(ctx); err != nil {
		slog.Debug("autoblock RulesWatcher: pub/sub unavailable, using poll-only mode", "error", err)
		pubsub.Close()
		return nil
	}

	slog.Debug("autoblock RulesWatcher: subscribed to push channel", "channel", channel)

	out := make(chan struct{}, 1)
	go func() {
		defer pubsub.Close()
		defer close(out)
		msgCh := pubsub.Channel()
		for {
			select {
			case <-ctx.Done():
				return
			case _, ok := <-msgCh:
				if !ok {
					return
				}
				// Non-blocking send: if a reload is already queued, skip duplicate.
				select {
				case out <- struct{}{}:
				default:
				}
			}
		}
	}()
	return out
}

// Poll performs a single reload cycle. Callable directly in tests.
func (w *RulesWatcher) Poll(ctx context.Context) error {
	raw, err := w.limiter.cfg.Redis.HGetAll(ctx, w.limiter.keys.rules()).Result()
	if err != nil {
		return err
	}

	dynamic := parseDynamicRules(raw)
	w.limiter.mergeRules(dynamic)

	if w.onReload != nil {
		w.onReload(dynamic)
	}
	return nil
}

func (w *RulesWatcher) handleError(err error) {
	slog.Warn("autoblock RulesWatcher poll error", "error", err)
	if w.onError != nil {
		w.onError(err)
	}
}

// ---------------------------------------------------------------------------
// Rule parsing from Redis hash
// ---------------------------------------------------------------------------

type redisRule struct {
	ID            string `json:"id"`
	Path          string `json:"path"`
	Limit         int    `json:"limit"`
	WindowSeconds int    `json:"window_seconds"`
	Algorithm     string `json:"algorithm"`
	PerUser       bool   `json:"per_user"`
	PerEndpoint   bool   `json:"per_endpoint"`
	Enabled       bool   `json:"enabled"`
}

func parseDynamicRules(raw map[string]string) []Rule {
	rules := make([]Rule, 0, len(raw))
	for _, jsonStr := range raw {
		var r redisRule
		if err := json.Unmarshal([]byte(jsonStr), &r); err != nil {
			continue
		}
		if !r.Enabled {
			continue
		}
		algo := Algorithm(r.Algorithm)
		if algo == "" {
			algo = AlgorithmHybrid
		}
		ws := r.WindowSeconds
		if ws <= 0 {
			ws = 60
		}
		rules = append(rules, Rule{
			Path:          r.Path,
			Limit:         r.Limit,
			WindowSeconds: ws,
			Algorithm:     algo,
			PerUser:       r.PerUser,
			PerEndpoint:   r.PerEndpoint,
		})
	}
	return rules
}

// ---------------------------------------------------------------------------
// Atomic rule swap on Limiter
// ---------------------------------------------------------------------------

// atomicRules is stored in Limiter to allow lock-free rule replacement.
type atomicRules struct {
	v atomic.Pointer[[]compiledRule]
}

func newAtomicRules(rules []compiledRule) *atomicRules {
	a := &atomicRules{}
	a.v.Store(&rules)
	return a
}

func (a *atomicRules) load() []compiledRule {
	return *a.v.Load()
}

func (a *atomicRules) store(rules []compiledRule) {
	a.v.Store(&rules)
}

// mergeRules replaces the live rule set with dynamic rules merged over the
// static config rules.  Thread-safe via atomic.Pointer.
func (l *Limiter) mergeRules(dynamic []Rule) {
	// Build set of dynamic ids to detect which static rules survive.
	dynIDs := make(map[string]bool, len(dynamic))
	for _, r := range dynamic {
		dynIDs[r.Path] = true // use Path as identifier (id not in Rule struct)
	}

	merged := make([]Rule, 0, len(dynamic)+len(l.cfg.Rules))
	merged = append(merged, dynamic...)
	for _, r := range l.cfg.Rules {
		if !dynIDs[r.Path] {
			merged = append(merged, r)
		}
	}

	compiled, err := compileRules(merged)
	if err != nil {
		slog.Warn("autoblock mergeRules compile error", "error", err)
		return
	}
	l.atomicRules.store(compiled)
}

File: packages/spring/src/main/java/io/autoblock/spring/core/RulesWatcher.java
package io.autoblock.spring.core;

import io.autoblock.spring.config.AutoBlockProperties;
import io.autoblock.spring.config.AutoBlockProperties.Algorithm;
import io.autoblock.spring.config.AutoBlockProperties.RuleProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.boot.json.BasicJsonParser;
import org.springframework.data.redis.connection.Message;
import org.springframework.data.redis.connection.MessageListener;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.listener.ChannelTopic;
import org.springframework.data.redis.listener.RedisMessageListenerContainer;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

/**
 * Polls {@code ab:{tenant}:rules:endpoint} (Redis hash) every 30 s AND
 * subscribes to the {@code ab:{tenant}:rules:changed} pub/sub channel so that
 * rule changes published by the AutoBlock engine are applied immediately
 * (&lt;1 s latency instead of up to 30 s).
 *
 * <p>Runs in a virtual thread ({@code Thread.ofVirtual()}) — no platform-thread
 * cost and no Spring {@code @Scheduled} dependency.
 *
 * <p>Wired as a bean by {@link io.autoblock.spring.AutoBlockAutoConfiguration}
 * when {@code autoblock.hot-reload.enabled=true} (default: true).
 */
public class RulesWatcher implements InitializingBean, DisposableBean {

    private static final Logger log = LoggerFactory.getLogger(RulesWatcher.class);
    private static final BasicJsonParser JSON = new BasicJsonParser();

    private final RateLimiter        limiter;
    private final RedisOps           redis;
    private final StringRedisTemplate redisTemplate;
    private final KeyBuilder         keys;
    private final Duration           interval;
    private final Consumer<List<RuleProperties>> onReload;

    private final AtomicBoolean running = new AtomicBoolean(false);
    private volatile Thread watcherThread;
    private volatile RedisMessageListenerContainer listenerContainer;

    /** Convenience constructor — no push subscriber (poll-only). */
    public RulesWatcher(
        RateLimiter limiter,
        RedisOps redis,
        AutoBlockProperties props
    ) {
        this(limiter, redis, null, props, Duration.ofSeconds(30), null);
    }

    public RulesWatcher(
        RateLimiter limiter,
        RedisOps redis,
        StringRedisTemplate redisTemplate,
        AutoBlockProperties props,
        Duration interval,
        Consumer<List<RuleProperties>> onReload
    ) {
        this.limiter       = limiter;
        this.redis         = redis;
        this.redisTemplate = redisTemplate;
        this.keys          = new KeyBuilder(props.tenant());
        this.interval      = interval;
        this.onReload      = onReload;
    }

    // ---- Lifecycle -------------------------------------------------------

    @Override
    public void afterPropertiesSet() {
        start();
    }

    @Override
    public void destroy() {
        stop();
    }

    public void start() {
        if (!running.compareAndSet(false, true)) return;  // idempotent

        // Immediate first poll (blocking, before traffic arrives)
        try {
            poll();
        } catch (Exception e) {
            log.warn("autoblock RulesWatcher initial poll failed: {}", e.getMessage());
        }

        watcherThread = Thread.ofVirtual()
            .name("autoblock-rules-watcher")
            .start(this::loop);

        // Subscribe for push-based invalidation (best-effort — fails gracefully).
        startPushSubscriber();
    }

    public void stop() {
        running.set(false);
        var t = watcherThread;
        if (t != null) t.interrupt();
        var c = listenerContainer;
        if (c != null) {
            try { c.stop(); } catch (Exception ignored) {}
        }
    }

    // ---- Core poll -------------------------------------------------------

    /**
     * Single poll cycle — publicly callable in tests.
     */
    public List<RuleProperties> poll() {
        var raw = redis.hGetAll(keys.rules());
        var dynamic = parseRules(raw);
        limiter.reloadRules(dynamic);
        if (onReload != null) onReload.accept(dynamic);
        log.debug("autoblock RulesWatcher reloaded {} dynamic rules", dynamic.size());
        return dynamic;
    }

    // ---- Background poll loop --------------------------------------------

    private void loop() {
        while (running.get()) {
            try {
                Thread.sleep(interval);
                poll();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            } catch (Exception e) {
                log.warn("autoblock RulesWatcher poll error: {}", e.getMessage());
            }
        }
    }

    // ---- Push subscriber -------------------------------------------------

    /**
     * Subscribes to {@code ab:{tenant}:rules:changed} via Spring Data Redis
     * pub/sub. On any message, triggers an immediate {@link #poll()}.
     * Gracefully skipped if {@code redisTemplate} is null.
     */
    private void startPushSubscriber() {
        if (redisTemplate == null) return;

        try {
            var factory = redisTemplate.getConnectionFactory();
            if (factory == null) return;

            MessageListener listener = (Message msg, byte[] pattern) -> {
                log.debug("autoblock RulesWatcher: push notification received, reloading immediately");
                try {
                    poll();
                } catch (Exception e) {
                    log.warn("autoblock RulesWatcher push-reload failed: {}", e.getMessage());
                }
            };

            var container = new RedisMessageListenerContainer();
            container.setConnectionFactory(factory);
            container.addMessageListener(listener, new ChannelTopic(keys.rulesChanged()));
            container.afterPropertiesSet();
            container.start();
            this.listenerContainer = container;

            log.debug("autoblock RulesWatcher: subscribed to push channel {}", keys.rulesChanged());
        } catch (Exception e) {
            log.debug("autoblock RulesWatcher: pub/sub unavailable ({}), poll-only mode", e.getMessage());
        }
    }

    // ---- Rule parsing ----------------------------------------------------

    private List<RuleProperties> parseRules(Map<Object, Object> raw) {
        var result = new ArrayList<RuleProperties>();
        for (var entry : raw.entrySet()) {
            try {
                var jsonStr = entry.getValue().toString();
                var r = JSON.parseMap(jsonStr);

                if (Boolean.FALSE.equals(r.get("enabled"))) continue;

                var path   = (String) r.getOrDefault("path", "");
                var limit  = toInt(r.getOrDefault("limit", 100));
                var window = toInt(r.getOrDefault("window_seconds", 60));
                var algo   = parseAlgorithm((String) r.getOrDefault("algorithm", "hybrid"));
                var perUser     = Boolean.TRUE.equals(r.get("per_user"));
                var perEndpoint = Boolean.TRUE.equals(r.get("per_endpoint"));

                if (path == null || path.isBlank() || limit <= 0) continue;

                result.add(new RuleProperties(path, limit, window, algo, perUser, perEndpoint));
            } catch (Exception e) {
                log.debug("autoblock RulesWatcher: skipping malformed rule entry: {}", e.getMessage());
            }
        }
        return result;
    }

    private static int toInt(Object v) {
        if (v instanceof Number n) return n.intValue();
        try { return Integer.parseInt(v.toString()); } catch (NumberFormatException e) { return 0; }
    }

    private static Algorithm parseAlgorithm(String raw) {
        if (raw == null) return Algorithm.HYBRID;
        return switch (raw.toLowerCase()) {
            case "sliding_window" -> Algorithm.SLIDING_WINDOW;
            case "token_bucket"   -> Algorithm.TOKEN_BUCKET;
            default               -> Algorithm.HYBRID;
        };
    }
}
